### PR TITLE
[Instrumentation.AspNetCore, Instrumentation.HttpClient] Add Enrichment instructions

### PR DESCRIPTION
**Pull request by ysolomchenko**
Fixes https://github.com/open-telemetry/opentelemetry-dotnet-contrib/issues/1762

## Changes

Add Enrichment instructions for .NET8.0 for
- Instrumentation.AspNetCore
- Instrumentation.HttpClient 

## Merge requirement checklist

* [x] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/CONTRIBUTING.md) guidelines followed (license requirements, nullable enabled, static analysis, etc.)
* [ ] Unit tests added/updated
* [ ] Appropriate `CHANGELOG.md` files updated for non-trivial changes
* [ ] Changes in public API reviewed (if applicable)

### CODE REVIEW

File: src/OpenTelemetry.Instrumentation.AspNetCore/README.md [hunk 1]
```diff
@@ -311,6 +311,52 @@ the library you can do so as follows:
 });
 ```
 
+#### Enriching ASP.NET Core Request Metrics
+
+ASP.NET Core supports enriching request metrics using `IHttpMetricsTagsFeature`.
+This feature allows you to add custom tags to metrics like `http.server.request.duration`,
+which records the duration of HTTP requests on the server.
+
+Here's an example of enriching the `http.server.request.duration` metric:
```
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.AspNetCore/README.md:320]: where do user specify that they intend to enrich `http.server.request.duration` metric, and nothing else?

File: src/OpenTelemetry.Instrumentation.AspNetCore/README.md [hunk 2]
```diff
@@ -311,6 +311,52 @@ the library you can do so as follows:
 });
 ```
 
+#### Enriching ASP.NET Core Request Metrics
+
+ASP.NET Core supports enriching request metrics using `IHttpMetricsTagsFeature`.
```
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.AspNetCore/README.md:316]: please call out that this is only applicable for .NET 8 and newer.

File: src/OpenTelemetry.Instrumentation.AspNetCore/README.md (hunk 1 above)
Comment (ysolomchenko) [reply src/OpenTelemetry.Instrumentation.AspNetCore/README.md:323]: Fixed in 99a4211340adb3c944e2380f9f010a72d6933dac

File: src/OpenTelemetry.Instrumentation.AspNetCore/README.md (hunk 2 above)
Comment (ysolomchenko) [reply src/OpenTelemetry.Instrumentation.AspNetCore/README.md:319]: Fixed in 99a4211340adb3c944e2380f9f010a72d6933dac

File: src/OpenTelemetry.Instrumentation.AspNetCore/README.md [hunk 3]
```diff
@@ -311,6 +311,75 @@ the library you can do so as follows:
...
+var builder = WebApplication.CreateBuilder(args);
+
+builder.Services.AddOpenTelemetry()
+    .WithMetrics(metricsBuilder =>
+    {
+        metricsBuilder
+            .SetResourceBuilder(ResourceBuilder.CreateDefault().AddService("MyService"))
+            .AddOtlpExporter()
+            .AddMeter("Microsoft.AspNetCore.Hosting", "Microsoft.AspNetCore.Server.Kestrel")
+            .AddView("http.server.request.duration",
```
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.AspNetCore/README.md:340]: do we need the View here?

File: src/OpenTelemetry.Instrumentation.AspNetCore/README.md [hunk 4]
```diff
@@ -311,6 +311,75 @@ the library you can do so as follows:
...
+
+var builder = WebApplication.CreateBuilder(args);
+
+builder.Services.AddOpenTelemetry()
+    .WithMetrics(metricsBuilder =>
+    {
+        metricsBuilder
+            .SetResourceBuilder(ResourceBuilder.CreateDefault().AddService("MyService"))
+            .AddOtlpExporter()
+            .AddMeter("Microsoft.AspNetCore.Hosting", "Microsoft.AspNetCore.Server.Kestrel")
```
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.AspNetCore/README.md:339]: this doc does not show the usage AddMeter with the name of meters passed earlier in the doc.. Lets stick to that pattern throughout the doc.

File: src/OpenTelemetry.Instrumentation.AspNetCore/README.md [hunk 5]
```diff
@@ -311,6 +311,75 @@ the library you can do so as follows:
...
+
+var app = builder.Build();
+
+// Middleware to enrich the request metric with marketing source
+app.Use(async (context, next) =>
+{
+    var tagsFeature = context.Features.Get<IHttpMetricsTagsFeature>();
+    if (tagsFeature != null)
+    {
+        var source = context.Request.Query["utm_medium"].ToString() switch
```
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.AspNetCore/README.md:356]: could we stick with a very minimal sample here? For Traces also we show enrichment from Context - can we use the same one here too?

File: src/OpenTelemetry.Instrumentation.AspNetCore/README.md [hunk 6]
```diff
@@ -311,6 +311,75 @@ the library you can do so as follows:
 });
 ```
 
+#### Enriching ASP.NET Core Request Metrics
+
+> [!IMPORTANT]
+> Only applicable for .NET 8 and newer.
+
+ASP.NET Core supports enriching request metrics using `IHttpMetricsTagsFeature`.
+This feature allows you to add custom tags to metrics like `http.server.request.duration`,
```
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.AspNetCore/README.md:320]: > like `http.server.request.duration`

I am not able to parse this part. What do we meant by "like" here? Is this enrichment applied to this specific metric, or any other metric that are similar to this or something else?  What if user creates custom metrics, will they also get enriched?

File: src/OpenTelemetry.Instrumentation.Http/README.md [hunk 7]
```diff
@@ -263,6 +263,69 @@ var tracerProvider = Sdk.CreateTracerProviderBuilder()
...
+sealed class EnrichmentHandler : DelegatingHandler
+{
+    protected override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+    {
+        HttpMetricsEnrichmentContext.AddCallback(request, static context =>
+        {
+            if (context.Response is not null) // Response is null when an exception occurs.
+            {
+                // Use any information available on the request or the response to emit custom tags.
+                string? value = context.Response.Headers.GetValues("Enrichment-Value").FirstOrDefault();
```
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.Http/README.md:299]: I strongly suggest to *not* show examples in the repo where Metric dimension are coming from HttpHeaders. We can stick with same example we use for Trace enrichment to showcase this capability.

File: src/OpenTelemetry.Instrumentation.Http/README.md [hunk 8]
```diff
@@ -263,6 +263,69 @@ var tracerProvider = Sdk.CreateTracerProviderBuilder()
     .Build();
 ```
 
+#### Enriching HttpClient Metrics
+
+> [!IMPORTANT]
+> Only applicable for .NET 8 and newer.
+
+Metrics enrichment in HttpClient allows adding custom tags to metrics, such as
```
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.Http/README.md:271]: same comment as the asp.net core one. It is unclear which metric(s) this custom tags are applied to.

**Review by cijothomas — CHANGES_REQUESTED** [review]
Thanks for improving the docs! I have requested changes to address the comments left.

File: src/OpenTelemetry.Instrumentation.AspNetCore/README.md (hunk 3 above)
Comment (ysolomchenko) [reply src/OpenTelemetry.Instrumentation.AspNetCore/README.md]: The AddView configuration for http.server.request.duration was removed

File: src/OpenTelemetry.Instrumentation.AspNetCore/README.md (hunk 5 above)
Comment (ysolomchenko) [reply src/OpenTelemetry.Instrumentation.AspNetCore/README.md]: The middleware has been simplified to focus on enriching metrics.

File: src/OpenTelemetry.Instrumentation.AspNetCore/README.md (hunk 6 above)
Comment (ysolomchenko) [reply src/OpenTelemetry.Instrumentation.AspNetCore/README.md]: IHttpMetricsTagsFeature applies specifically to metrics generated by the ASP.NET Core, but does not apply to custom user-defined metrics automatically.

File: src/OpenTelemetry.Instrumentation.AspNetCore/README.md (hunk 4 above)
Comment (ysolomchenko) [reply src/OpenTelemetry.Instrumentation.AspNetCore/README.md]: Changes were made

File: src/OpenTelemetry.Instrumentation.AspNetCore/README.md [hunk 9]
```diff
@@ -311,6 +311,56 @@ the library you can do so as follows:
...
+var app = builder.Build();
+
+// Middleware to enrich the request metric with a custom tag
+app.Use(async (context, next) =>
+{
+    var tagsFeature = context.Features.Get<IHttpMetricsTagsFeature>();
+    if (tagsFeature != null)
+    {
+        // Add a custom tag based on the "utm_medium" query parameter
+        var source = context.Request.Query["utm_medium"].ToString();
```
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.AspNetCore/README.md:345]: Its best if we show good practices in the docs - can we remove the string allocation in the hot path?

File: src/OpenTelemetry.Instrumentation.AspNetCore/README.md (hunk 1 above)
Comment (cijothomas) [reply src/OpenTelemetry.Instrumentation.AspNetCore/README.md:322]: I am still confused. Where does user specify that they intend to enrich "http.server.request.duration" metric, and nothing else?
The example just shows how to get `IHttpMetricsTagsFeature` and add tags to it. But don't see any place where user selects which metric is this applied to.

File: src/OpenTelemetry.Instrumentation.AspNetCore/README.md [hunk 10]
```diff
@@ -311,6 +311,83 @@ the library you can do so as follows:
...
+app.MapGet("/", () => "Hello OpenTelemetry!");
+
+app.Run();
+```
+
+In this example:
+
+* Automatic Metric Enrichment: The IHttpMetricsTagsFeature enriches all relevant
+  HTTP metrics(like http.server.request.duration) by default. There is no need to
+  explicitly specify the metric you want to enrich.
```
Comment (Kielek) [inline src/OpenTelemetry.Instrumentation.AspNetCore/README.md:382]:
```suggestion
* Automatic Metric Enrichment: The `IHttpMetricsTagsFeature` enriches all relevant
  HTTP metrics (like `http.server.request.duration`) by default. There is no need to
  explicitly specify the metric you want to enrich.
```

File: src/OpenTelemetry.Instrumentation.AspNetCore/README.md [hunk 11]
```diff
@@ -311,6 +311,83 @@ the library you can do so as follows:
...
+* Context-Specific Tags: In the middleware, custom tags such as utm_medium are
+  added to the request's metrics context. These tags will automatically appear in
+  the exported data for any related metrics.
+* Customizing Exported Metrics: To focus on enriching or exporting a specific
+  metric (e.g., http.server.request.duration), you can control this at the exporting
+  stage, either by filtering specific metrics in the exporter or by customizing
+  the telemetry pipeline configuration.
```
Comment (Kielek) [inline src/OpenTelemetry.Instrumentation.AspNetCore/README.md:389]:
```suggestion
* Customizing Exported Metrics: To focus on enriching or exporting a specific
  metric (e.g. `http.server.request.duration`), you can control this at the exporting
  stage, either by filtering specific metrics in the exporter or by customizing
  the telemetry pipeline configuration.
```

File: src/OpenTelemetry.Instrumentation.Http/README.md [hunk 12]
```diff
@@ -263,6 +263,65 @@ var tracerProvider = Sdk.CreateTracerProviderBuilder()
     .Build();
 ```
 
+#### Enriching HttpClient Metrics
+
+> [!IMPORTANT]
+> Only applicable for .NET 8 and newer.
+
+Metrics enrichment in HttpClient allows adding custom tags to metrics.
+ This is especially useful for categorizing metrics in dashboards or alerts.
```
Comment (Kielek) [inline src/OpenTelemetry.Instrumentation.Http/README.md:272]:
```suggestion
This is especially useful for categorizing metrics in dashboards or alerts.
```

**Pull request closed without merging**